sa/saro.go

```diff
@@ -0,0 +1,1437 @@
+// purpose of rate limiting using a range that is inclusive on the lower end and
+// exclusive at the higher end. If ip is an IPv4 address, it returns that address,
+// plus the one immediately higher than it. If ip is an IPv6 address, it applies
+// a /48 mask to it and returns the lowest IP in the resulting network, and the
+// first IP outside of the resulting network.
+func ipRange(ip net.IP) (net.IP, net.IP) {
+	ip = ip.To16()
+	// For IPv6, match on a certain subnet range, since one person can commonly
+	// have an entire /48 to themselves.
+	maskLength := 48
+	// For IPv4 addresses, do a match on exact address, so begin = ip and end =
+	// next higher IP.
+	if ip.To4() != nil {
+		maskLength = 128
+	}
+
+	mask := net.CIDRMask(maskLength, 128)
+	begin := ip.Mask(mask)
+	end := incrementIP(begin, maskLength)
+
+	return begin, end
+}
+
+// CountRegistrationsByIP returns the number of registrations created in the
+// time range for a single IP address.
+func (ssa *SQLStorageAuthorityRO) CountRegistrationsByIP(ctx context.Context, req *sapb.CountRegistrationsByIPRequest) (*sapb.Count, error) {
+	if len(req.Ip) == 0 || req.Range.Earliest == 0 || req.Range.Latest == 0 {
+		return nil, errIncompleteRequest
+	}
+
+	var count int64
+	err := ssa.dbReadOnlyMap.WithContext(ctx).SelectOne(
+		&count,
+		`SELECT COUNT(1) FROM registrations
+		 WHERE
+		 initialIP = :ip AND
+		 :earliest < createdAt AND
+		 createdAt <= :latest`,
+		map[string]interface{}{
+			"ip":       req.Ip,
+			"earliest": time.Unix(0, req.Range.Earliest),
+			"latest":   time.Unix(0, req.Range.Latest),
+		})
+	if err != nil {
+		return &sapb.Count{Count: -1}, err
+	}
+	return &sapb.Count{Count: count}, nil
+}
+
+func (ssa *SQLStorageAuthority) CountRegistrationsByIP(ctx context.Context, req *sapb.CountRegistrationsByIPRequest) (*sapb.Count, error) {
+	return ssa.SQLStorageAuthorityRO.CountRegistrationsByIP(ctx, req)
+}
+
+// CountRegistrationsByIPRange returns the number of registrations created in
+// the time range in an IP range. For IPv4 addresses, that range is limited to
+// the single IP. For IPv6 addresses, that range is a /48, since it's not
+// uncommon for one person to have a /48 to themselves.
+func (ssa *SQLStorageAuthorityRO) CountRegistrationsByIPRange(ctx context.Context, req *sapb.CountRegistrationsByIPRequest) (*sapb.Count, error) {
+	if len(req.Ip) == 0 || req.Range.Earliest == 0 || req.Range.Latest == 0 {
+		return nil, errIncompleteRequest
+	}
+
+	var count int64
+	beginIP, endIP := ipRange(req.Ip)
+	err := ssa.dbReadOnlyMap.WithContext(ctx).SelectOne(
+		&count,
+		`SELECT COUNT(1) FROM registrations
+		 WHERE
+		 :beginIP <= initialIP AND
+		 initialIP < :endIP AND
+		 :earliest < createdAt AND
+		 createdAt <= :latest`,
+		map[string]interface{}{
+			"earliest": time.Unix(0, req.Range.Earliest),
+			"latest":   time.Unix(0, req.Range.Latest),
+			"beginIP":  beginIP,
+			"endIP":    endIP,
+		})
+	if err != nil {
+		return &sapb.Count{Count: -1}, err
```

Comment:
First, I understand that this is all coming over from `sa.go`, but I couldn't help but notice that we return `count: -1, err` (if `err != nil`) from just under half of SA methods which use `SELECT COUNT(1) FROM...`. The callers of these methods don't seem to depend on this behavior. Is there a benefit that I'm missing? If so, do we want to standardize on this?

Reply:
I hadn't even noticed that. I think that digging into it is definitely a good idea, but also definitely not in this change. In big refactors like this, I like to be able to confidently say "no code changed, it only moved". Want to file a bug for it?

Reply:
Completely agreed there. Opened #6494 for tracking.